c/src/ml-api-service-extension.c

```diff
@@ -0,0 +1,1238 @@
+    if (status != ML_ERROR_NONE) {
+      _ml_error_report
+          ("Failed to parse configuration file, it should have valid model path.");
+      goto error;
+    }
+
+    ml_option_set (option, "models", paths, g_free);
+  } else {
+    status = ML_ERROR_INVALID_PARAMETER;
+    _ml_error_report
+        ("Failed to parse configuration file, cannot get the model path.");
+    goto error;
+  }
+
+  if (json_object_has_member (single, "framework")) {
+    const gchar *fw = json_object_get_string_member (single, "framework");
+
+    if (STR_IS_VALID (fw))
+      ml_option_set (option, "framework_name", g_strdup (fw), g_free);
+  }
+
+  if (json_object_has_member (single, "input_info")) {
+    JsonNode *info_node = json_object_get_member (single, "input_info");
+    ml_tensors_info_h in_info;
+
+    status = _ml_extension_conf_parse_tensors_info (info_node, &in_info);
+    if (status != ML_ERROR_NONE) {
+      _ml_error_report
+          ("Failed to parse configuration file, cannot parse input information.");
+      goto error;
+    }
+
+    ml_option_set (option, "input_info", in_info,
+        _ml_extension_destroy_tensors_info);
+  }
+
+  if (json_object_has_member (single, "output_info")) {
+    JsonNode *info_node = json_object_get_member (single, "output_info");
+    ml_tensors_info_h out_info;
+
+    status = _ml_extension_conf_parse_tensors_info (info_node, &out_info);
+    if (status != ML_ERROR_NONE) {
+      _ml_error_report
+          ("Failed to parse configuration file, cannot parse output information.");
+      goto error;
+    }
+
+    ml_option_set (option, "output_info", out_info,
+        _ml_extension_destroy_tensors_info);
+  }
+
+  if (json_object_has_member (single, "custom")) {
+    const gchar *custom = json_object_get_string_member (single, "custom");
+
+    if (STR_IS_VALID (custom))
+      ml_option_set (option, "custom", g_strdup (custom), g_free);
+  }
+
+error:
+  if (status == ML_ERROR_NONE)
+    status = ml_single_open_with_option (&ext->single, option);
+
+  ml_option_destroy (option);
+  return status;
+}
+
+/**
+ * @brief Internal function to parse the node info in pipeline.
+ */
+static int
+_ml_extension_conf_parse_pipeline_node (ml_service_s * mls, JsonNode * node,
+    ml_extension_node_type_e type)
+{
+  ml_extension_s *ext = (ml_extension_s *) mls->priv;
+  JsonArray *array = NULL;
+  JsonObject *object;
+  guint i, n;
+  int status;
+
+  n = 1;
+  if (JSON_NODE_HOLDS_ARRAY (node)) {
+    array = json_node_get_array (node);
+    n = json_array_get_length (array);
+  }
+
+  for (i = 0; i < n; i++) {
+    const gchar *name = NULL;
+    ml_extension_node_info_s *node_info;
+
+    if (array)
+      object = json_array_get_object_element (array, i);
+    else
+      object = json_node_get_object (node);
+
+    if (json_object_has_member (object, "name"))
+      name = json_object_get_string_member (object, "name");
+
+    node_info = _ml_extension_node_info_new (mls, name, type);
```

Comment:
node_info is a local pointer variable. but the memory is not being freed

Reply:
The node-info is inserted into hashtable in ml-service handle. If failed to create ml-service handle, the hashtable will be destroyed.

Reply:
Thank you.